doc/lsst.daf.butler/formatters.rst

```diff
@@ -0,0 +1,316 @@
+On read the the storage class used to read the file can be different from the storage class expected to be returned by `~lsst.daf.butler.Datastore`.
+This happens if a composite was written but a component from that composite is being read.
+
+File Extensions
+^^^^^^^^^^^^^^^
+
+Each formatter that reads or writes a file must declare the file extensions that it supports.
+For a formatter that supports a single extension this is most easily achieved by setting the class property `~lsst.daf.butler.Formatter.extension` to that extension.
+In some scenarios a formatter might support multiple formats that are controlled by write parameters.
+In this case the formatter should assign a frozen set to the `~lsst.daf.butler.Formatter.supportedExtensions` class property.
+It is then required that the class implement an instance property for ``extension`` that returns the extension that will be used by this formatter for writing the current dataset.
+
+File vs Bytes
+^^^^^^^^^^^^^
+
+Some datastores can stream bytes from remote storage systems and do not require that a local file is created before the Python object can be created.
+To support this use case an implementer can implement `~lsst.daf.butler.Formatter.fromBytes()` for reading in from a datastore and `~lsst.daf.butler.Formatter.toBytes()` for serializing to a datastore.
+If a formatter raises `NotImplementedError` when these byte-like methods are called the datastore will default to using the `~lsst.daf.butler.Formatter.read()` and `~lsst.daf.butler.Formatter.write()` methods making use of local temporary files.
+
+.. warning::
+
+  This interface has some rough edges since it is not yet possible for the formatter to optionally support bytes directly based on the amount of data involved.
+  Even though bytes may be more efficient for small or medium-sized datasets, in some cases with significant datasets the memory overhead of multiple copies may be excessive and a temporary file would be more prudent.
+  Neither datastore nor the formatter can opt out of using bytes on a per-dataset basis.
+
+FileFormatter Subclass
+^^^^^^^^^^^^^^^^^^^^^^
+
+For many file-based formatter implementations a subclass of `~lsst.daf.butler.Formatter` can be used that has a much simplified interface.
+`~lsst.daf.butler.formatters.file.FileFormatter` allows a formatter implementation to be written using two methods: `~lsst.daf.butler.formatters.file.FileFormatter._readFile()` takes a local path to the file system and the expected Python type, and `~lsst.daf.butler.formatters.file.FileFormatter._writeFile()` takes the in-memory object to be serialized.
+
+Composites are not handled by `~lsst.daf.butler.formatters.file.FileFormatter`.
+
+.. note::
+  I'm not sure I understand why _writeFile() doesn't also take the path rather than requiring FileDescriptor to be used.
+  It's inconsistent with _readFile that does take the local path.
+  It's not much of a simplification as things stand.
+  Need to revisit that.
```

Comment:
Should this be ticketed?

Reply:
Step 1 was getting feedback from this review...

Reply:
I do think some work here is worthwhile.  Some related considerations:
 - Some prominent file-based formatters (`FiitsExposureFormatter` in obs_base, at least) can't use `FileFormatter` and have to inherit from `Formatter` directly.  I don't recall why, or even if it's still a problem.
 - I don't see why `FileFormatter` (rather than base `Formatter`, or even code that calls it) is responsible for updating file extensions in `write`.  I also have hard time keeping track of whether one is supposed to modify `self.fileDescriptor` in place, or whether that might cause bad side-effects, and I'd really like to keep `Formatter` authors from having to deal with any of that.

It'd certainly be much easier to write `Formatters` if there was a simplified interface that _only_ dealt with local file paths, took care of extensions, and didn't require the author to learn about `FileDescriptor` at all.

Reply:
I'll create a ticket for this. The urgency for it partly depends on how soon other people are going to be writing their own formatters.

Reply:
DM-26658